### PR TITLE
Allow docvalues-only search on geo_shape

### PR DESCRIPTION
**Pull request by iverase**
Similar to https://github.com/elastic/elasticsearch/pull/82409, this allows searching on a geo_shape field type when the field is not indexed (index: false) but just doc values are enabled.

### CODE REVIEW

File: docs/reference/mapping/types/geo-shape.asciidoc [hunk 1]
```diff
@@ -70,6 +70,14 @@ and reject the whole document.
 |`coerce` |If `true` unclosed linear rings in polygons will be automatically closed.
 | `false`
 
+|`index` |Should the field be quickly searchable? Accepts `true` (default) and `false`.
+Fields that only have <<doc-values,`doc_values`>> enabled can still be queried, albeit slower.
+| `true`
+
+|`doc_values` |Should the field be stored on disk in a column-stride fashion,
```
Comment (craigtaverner) [inline docs/reference/mapping/types/geo-shape.asciidoc:77]: Can both `index` and `doc_values` be set to `false`?

**Review by craigtaverner — APPROVED** [review]
LGTM. I did have comments, but I think they answered themselves.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java [hunk 2]
```diff
@@ -197,17 +197,23 @@ public String typeName() {
...
-                final Query queryDocValues = new LatLonShapeDocValuesQuery(fieldName, relation.getLuceneRelation(), geometries);
-                query = new IndexOrDocValuesQuery(query, queryDocValues);
+            Query query;
+            if (isIndexed()) {
+                query = LatLonShape.newGeometryQuery(fieldName, relation.getLuceneRelation(), geometries);
+                if (hasDocValues()) {
+                    final Query queryDocValues = new LatLonShapeDocValuesQuery(fieldName, relation.getLuceneRelation(), geometries);
+                    query = new IndexOrDocValuesQuery(query, queryDocValues);
+                }
+            } else {
```
[the remark this answers is not in the document]
Comment (iverase) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java:215]: This case is handle by `failIfNotIndexedNorDocValuesFallback(context);`

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/spatial/110_geo_shape_parameters.yml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                  lon: 4
+                bottom_right:
+                  lat: 50
+                  lon: 5
+
+  - match: {error.type: search_phase_execution_exception}
+  - match: {error.reason: "all shards failed"}
+  - match: {error.phase: query}
+  - match: {error.failed_shards.0.reason.type:  query_shard_exception}
+  - match: {error.failed_shards.0.reason.reason: "failed to create query: Cannot search on field [no_doc_values_no_index] since it is not indexed nor has doc values."}
```
[the remark this answers is not in the document]
Comment (iverase) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/spatial/110_geo_shape_parameters.yml:127]: This is something generic and I don't think we should explicitly document it on this field mapping.

**Pull request merged**